### PR TITLE
feat(sdk): Live updates for pinned events timeline with replacements and redactions

### PR DESCRIPTION
**Pull request by jmartinesp**
This is a follow-up PR for https://github.com/matrix-org/matrix-rust-sdk/pull/3870.

## Changes

- Make `TimelineEventBuilder` process synced events for `TimelineFocus::PinnedEvents` too.
- Use the relationship cache added in https://github.com/matrix-org/matrix-rust-sdk/pull/3870 to load pinned events and other events related to them so their final timeline event results contain the changes those would add (edits, redactions, reactions, etc.).
- Make `Room::event_with_context` also save its results to the cache.
- Add new tests for pinned events and related events.
- Fix the pinned events benchmark in `room_bench`.

The end result of this is the pinned events timeline now reacts to edits and redactions, and those values are kept even when the event list reloads.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs [hunk 1]
```diff
@@ -63,20 +63,24 @@ impl PinnedEventsLoader {
...
 
         let new_events = join_all(pinned_event_ids.into_iter().map(|event_id| {
+            let semaphore = semaphore.clone();
             let provider = self.room.clone();
             async move {
-                match provider.load_event(&event_id, request_config).await {
+                if let Err(err) = semaphore.acquire().await {
+                    warn!("error when loading pinned event: {err}");
+                    return None;
+                }
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:83]: Pretty sure this is a no-op when the semaphore could be acquired, because the guard (in the `Ok()` value) is not bound, so it's immediately dropped. Can you add a test for this? (I would even recommend making a separate PR for this feature, since it seems unrelated to the current PR's purpose, and could land separately.)

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 2]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
                 continue;
             };
 
+            // Handle and cache events and relations
+            let mut relations = self.relations.write().await;
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs]: Unfortunately, this is not the right place to add this; I suppose this should be in the same place where I inserted events into `Self::events`. Follow the breadcrumbs :)

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 3]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
...
+            }
+            drop(relations);
+
+            let mut events = self.events.write().await;
+            for ev in &joined_room_update.timeline.events {
+                if let Some(ev_id) = ev.event_id() {
+                    events.insert(ev_id, (room_id.clone(), ev.clone()));
+                }
+            }
+            drop(events);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs]: This can entirely be removed, because it's already done a few layers below.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 4]
```diff
@@ -501,6 +558,29 @@ impl RoomEventCache {
         None
     }
 
+    /// Try to find an event by id in this room, along with all relations.
+    pub async fn event_with_relations(&self, event_id: &EventId) -> Vec<SyncTimelineEvent> {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:583]: Can you return an `(SyncTimelineEvent, Vec<SyncTimelineEvent>)` instead, so it's clear where the target event is? (Right now there's nothing either in the signature or the doc-comment that makes it obvious it's the first one.)

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 5]
```diff
@@ -501,6 +558,29 @@ impl RoomEventCache {
...
+        let relations = self.inner.relations.read().await;
+        let relation_ids = relations.get(event_id);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:585]: Can you inline this where it's used, instead? It's not useful to preload it, if the event couldn't be found.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 6]
```diff
@@ -312,6 +316,7 @@ impl EventCache {
 }
 
 type AllEventsMap = Arc<RwLock<BTreeMap<OwnedEventId, (OwnedRoomId, SyncTimelineEvent)>>>;
+type RelationsMap = Arc<RwLock<BTreeMap<OwnedEventId, BTreeSet<OwnedEventId>>>>;
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:319]: There are two locks for data that's touched around the same time; I'd suggest grouping the two fields inside a data structure, and using a single lock, to avoid having to think about lock ordering.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 7]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
                 continue;
             };
 
+            // Handle and cache events and relations
+            let mut relations = self.relations.write().await;
+            for ev in &joined_room_update.timeline.events {
+                if let Ok(AnySyncTimelineEvent::MessageLike(ev)) = ev.event.deserialize() {
+                    if let Some(AnyMessageLikeEventContent::RoomMessage(c)) = ev.original_content()
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:411]: Can you use a `match` statement so we get a single call to `ev.original_content()`?

Also, do you want to keep reactions too?

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 8]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
...
+                    {
+                        if let Some(Relation::Replacement(relation)) = c.relates_to {
+                            if let Some(replacements) = relations.get_mut(&relation.event_id) {
+                                replacements.insert(ev.event_id().to_owned());
+                            } else {
+                                relations.insert(
+                                    relation.event_id,
+                                    BTreeSet::from_iter(vec![ev.event_id().to_owned()]),
+                                );
+                            }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs]: You can probably use `relations.entry(relation.event_id.clone()).or_default().insert(…)` to make this a bit shorter to read.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 9]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
...
+                        if let Some(redacted_ev_id) = c.redacts {
+                            let original_ev_id = redacted_ev_id;
+                            if let Some(replacements) = relations.get_mut(&original_ev_id) {
+                                replacements.insert(ev.event_id().to_owned());
+                            } else {
+                                relations.insert(
+                                    original_ev_id,
+                                    BTreeSet::from_iter(vec![ev.event_id().to_owned()]),
+                                );
+                            }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs]: Same remark for `.entry()`.

File: crates/matrix-sdk/src/room/mod.rs [hunk 10]
```diff
@@ -445,6 +445,21 @@ impl Room {
         )
         .await?;
 
+        // Save the loaded events into the event cache, if it's set up.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/room/mod.rs:449]: Nice! Can you test this please? :)

Also consider adding a `RoomEventCache::save_events` method, so that we can save all the events at once without giving and releasing and giving and releasing etc. the lock.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs [hunk 11]
```diff
@@ -74,6 +77,7 @@ impl PinnedEventsLoader {
             let provider = self.room.clone();
             async move {
                 match provider.load_event(&event_id, request_config).await {
+                match provider.load_event_with_relations(&event_id, request_config).await {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs]: uh?

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs [hunk 12]
```diff
@@ -124,22 +130,23 @@ pub trait PinnedEventsRoom: SendOutsideWasm + SyncOutsideWasm {
...
+                return Ok(cached_event_with_relations);
             }
         }
 
         debug!("Loading pinned event {event_id} from HS");
-        self.event(event_id, request_config)
+        self.event_with_context(event_id, true, uint!(0), request_config)
             .await
-            .map(|e| e.into())
+            .map(|e| if let Some(ev) = e.event { vec![ev.into()] } else { Vec::new() })
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:169]: nit: use a better name of `e`, and what's happening to related events here?

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 13]
```diff
@@ -396,6 +404,49 @@ impl EventCacheInner {
...
+                                relations.insert(
+                                    relation.event_id,
+                                    BTreeSet::from_iter(vec![ev.event_id().to_owned()]),
+                                );
+                            }
+                        }
+                    } else if let Some(AnyMessageLikeEventContent::RoomRedaction(c)) =
+                        ev.original_content()
+                    {
+                        if let Some(redacted_ev_id) = c.redacts {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:443]: It's optional because in previous versions than v11, the redacts field was somewhere else, if I remember correctly (in the event itself, not the content). Can you handle that case too?

File: crates/matrix-sdk-ui/tests/integration/main.rs [hunk 14]
```diff
@@ -84,7 +84,7 @@ async fn mock_context(
 }
 
 /// Mocks the /event endpoint
-#[allow(clippy::too_many_arguments)] // clippy you've got such a fixed mindset
+#[allow(clippy::too_many_arguments, dead_code)] // clippy you've got such a fixed mindset
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/main.rs:87]: Why the `dead_code`? If it's unused, let's get rid of it.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 15]
```diff
@@ -355,20 +599,56 @@ impl TestHelper {
...
         let mut joined_room_builder = JoinedRoomBuilder::new(&self.room_id);
-        for (id, txt, add_to_timeline) in text_messages {
-            let event_id: OwnedEventId = id.try_into().unwrap();
-            let f = EventFactory::new().room(&self.room_id);
-            let event_builder = f.text_msg(txt).event_id(&event_id).sender(*BOB);
-            mock_event(&self.server, &self.room_id, &event_id, event_builder.into_timeline()).await;
+        for (message_type, add_to_timeline) in text_messages {
+            let (event_id, timeline_item) = match message_type {
+                MessageType::New { id, txt } => {
+                    let f = EventFactory::new().room(&self.room_id);
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:609]: nit: reuse the same `EventFactory` instead of creating a new one each time.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 16]
```diff
@@ -355,20 +599,56 @@ impl TestHelper {
...
+                    let event_builder = f
+                        .text_msg(txt)
+                        .event_id(&id)
+                        .sender(*BOB)
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:613]: Since the sender is always `BOB`, you can call `f.sender(*BOB)` on the `EventFactory` itself.

File: crates/matrix-sdk/src/test_utils/events.rs [hunk 17]
```diff
@@ -213,6 +219,17 @@ impl EventFactory {
...
+    /// Create a replacement message for an existing one given its id.
+    pub fn replacement_msg(
+        &self,
+        new_content: impl Into<String>,
+        original_event_id: &EventId,
+    ) -> EventBuilder<RoomMessageEventContent> {
+        let msg_content = RoomMessageEventContent::text_plain(new_content.into());
+        let builder = self.event(msg_content.clone());
+        builder.replaces(original_event_id, msg_content)
+    }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/test_utils/events.rs:231]: Is this function needed? It's just a shortcut to other functions which can be used directly, and it would avoid adding too much abstraction in the tests themselves.

File: crates/matrix-sdk/src/test_utils/events.rs [hunk 18]
```diff
@@ -115,6 +115,12 @@ impl EventBuilder<RoomMessageEventContent> {
             Some(Relation::Reply { in_reply_to: InReplyTo::new(event_id.to_owned()) });
         self
     }
+
+    pub fn replaces(mut self, event_id: &EventId, new_content: RoomMessageEventContent) -> Self {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/test_utils/events.rs:119]: Can you add a doc-comment here please?

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 19]
```diff
@@ -402,3 +682,28 @@ impl TestHelper {
         self.client.sync_once(self.sync_settings.clone()).await
     }
 }
+
+enum MessageType {
+    New { id: OwnedEventId, txt: String },
+    Edit { id: OwnedEventId, old_id: OwnedEventId, txt: String },
+    Redaction { id: OwnedEventId, old_id: OwnedEventId },
+}
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:690]: Instead of introducing all of this, could the caller make use of an `EventFactory` immediately, and not pass a `MessageType` etc.? This would make the tests much more straightforward to read…

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 20]
```diff
@@ -355,20 +599,56 @@ impl TestHelper {
...
         pinned_event_ids: Option<Vec<&str>>,
     ) -> Result<SyncResponse, matrix_sdk::Error> {
         let mut joined_room_builder = JoinedRoomBuilder::new(&self.room_id);
-        for (id, txt, add_to_timeline) in text_messages {
-            let event_id: OwnedEventId = id.try_into().unwrap();
-            let f = EventFactory::new().room(&self.room_id);
-            let event_builder = f.text_msg(txt).event_id(&event_id).sender(*BOB);
-            mock_event(&self.server, &self.room_id, &event_id, event_builder.into_timeline()).await;
+        for (message_type, add_to_timeline) in text_messages {
+            let (event_id, timeline_item) = match message_type {
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:607]: nit: not `timeline_item`, but `timeline_event`, I think; item should be reserved to the UI Timeline object :)

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 21]
```diff
@@ -319,6 +329,240 @@ async fn test_pinned_timeline_with_no_pinned_event_ids_is_just_empty() {
...
+    let (items, mut timeline_stream) = timeline.subscribe().await;
+
+    assert_eq!(items.len(), 1 + 1); // event item + a day divider
+    assert!(items[0].is_day_divider());
+    assert_eq!(items[1].as_event().unwrap().content().as_message().unwrap().body(), "in the end");
+    assert_pending!(timeline_stream);
+    test_helper.server.reset().await;
+
+    // Load new pinned event contents from sync, where $2 is and edit on $1
+    let _ = Box::pin(
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:370]: Why is the `Box::pin` needed? Can you do without?

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs [hunk 22]
```diff
@@ -319,6 +329,240 @@ async fn test_pinned_timeline_with_no_pinned_event_ids_is_just_empty() {
...
+        test_helper.setup_sync_response(vec![(edit("$2", "$1", "edited message!"), true)], None),
+    )
+    .await;
+
+    // The list is reloaded, so it's reset
+    assert_matches!(timeline_stream.next().await.unwrap(), VectorDiff::Clear);
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:376]: I think there's an `assert_next` that exists and could replace most of these `assert_matches` invokes here and below, and in the other test(s) too?

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 23]
```diff
@@ -496,6 +504,18 @@ impl EventCacheInner {
...
+    async fn insert_or_update_relationship(&self, orig_ev_id: &EventId, related_ev_id: &EventId) {
+        let mut relations = self.relations.write().await;
+
+        if let Some(replacements) = relations.get_mut(orig_ev_id) {
+            replacements.insert(related_ev_id.to_owned());
+        } else {
+            relations
+                .insert(orig_ev_id.to_owned(), BTreeSet::from_iter(vec![related_ev_id.to_owned()]));
+        }
+        drop(relations);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs]: nit: no need to drop explicitly, it'll happen automatically at the end of the function's body.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 24]
```diff
@@ -496,6 +504,18 @@ impl EventCacheInner {
             }
         }
     }
+
+    async fn insert_or_update_relationship(&self, orig_ev_id: &EventId, related_ev_id: &EventId) {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/mod.rs:509]: nit: please fully spell out `event_id` for consistency, and `original` too

(I for one would call these two parameters `original` and `related`, because including the parameter's type is spurious, since the parameter type is next to the variable name, but that's my preference :))

File: benchmarks/benches/room_bench.rs [hunk 25]
```diff
@@ -186,6 +187,8 @@ pub fn load_pinned_events_benchmark(c: &mut Criterion) {
 
     group.bench_function(BenchmarkId::new("load_pinned_events", name), |b| {
         b.to_async(&runtime).iter(|| async {
+            client.event_cache().subscribe().unwrap();
```
Comment (bnjbvr) [inline benchmarks/benches/room_bench.rs:190]: Is there really a need to do it for *each* iteration of the benchmark? I don't think so, it should be subscribed to once and for all.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Looks like you're on the right track! I'd like a few test changes, and the main logic in the `EventCache` itself isn't tested independently. Please add some tests for that too, since it's a different layer than the timeline.

Also, please reword the PR title to follow our conventional commit guidelines :)
Thanks!

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs (hunk 1 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs:83]: I have no idea how to test that, to be honest. Since there is no timeout I think the execution will just get stuck there until a new permit is available.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 2 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs]: I think I'll have to do that in a separate method then? I don't think I can have 2 `RwLockWriteGuards` at the same time, one for `all_events` and another one for `relations`, so I don't think I'll be able to add this on `append_events_locked_impl`.

Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs]: Oh, I just saw https://github.com/matrix-org/matrix-rust-sdk/pull/3840#discussion_r1716841925.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs (hunk 11 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs]: I think it was a rebase hiccup, the code is not there later 😅 .

File: crates/matrix-sdk-ui/tests/integration/main.rs (hunk 14 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/main.rs:87]: It's a helper for mocking data, so I expected it to be useful in the future. I can remove it though.

File: crates/matrix-sdk/src/test_utils/events.rs (hunk 17 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/test_utils/events.rs:231]: It's a helper, so I expect it to contain boilerplate to avoid repetitive code. Should I remove the `redaction` fn above too?

File: crates/matrix-sdk/src/test_utils/events.rs (hunk 18 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/test_utils/events.rs:119]: Oh, sure. I though I had handled all the pub fns but I missed this one at least.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 19 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:690]: That means inlining the whole `setup_sync_response` code for every sync. I honestly think that's way less readable, and it's the main reason why I abstracted it inside `TestHelper` in a previous PR.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 21 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:370]: Oops, leftover from one of my tests to check why the tests were stuck.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 23 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs]: Yep, copy&paste issue. It's gone now.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 24 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs:509]: I'm removing this code so it shouldn't be needed anymore.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 22 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:376]: For some reason changing `assert_matches!(timeline_stream.next().await.unwrap(), *)` to `assert_next_matches!(timeline_stream, *)` makes one of the tests fail with:

> assertion failed: stream is not ready, it's pending

Maybe I need to add a delay before? It's quite weird.

Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:376]: Oh, it's because under the hood it calls `Stream::next_now_or_never`... That seems dangerous. I don't see any similar function for doing it while awaiting for the next item.

File: benchmarks/benches/room_bench.rs (hunk 25 above)
Comment (jmartinesp) [reply benchmarks/benches/room_bench.rs:190]: Maybe I can wrap `Client` in an `Arc` and use `runtime.spawn_blocking` to run that code? Is there a better solution?

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 7 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs:411]: Reactions weren't planned for the pined events feature, but I can try to add them too.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 13 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs:443]: So... this seems way more difficult than expected. In the `EventCache` I have no way to access the `RoomBaseInfo` to check the room version.

Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs:443]: Maybe I can try not relying on the version id and just checking the event: if the redaction is in the event itself, use it, otherwise, check for the redaction inside the contents.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 19 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:681]: Can you pass the `SyncTimelineEvent`s or `TimelineEvent`s to `setup_sync_responses`?

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 22 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs:375]: We need to add a timeout then, otherwise if the stream item doesn't come immediately, then the test will be timing out with `cargo-nextest`'s delay, which is a minute or longer.

File: crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs (hunk 1 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/src/timeline/pinned_events_loader.rs]: I replaced this with a buffered stream now. It seems to work just as fine and there is no need for the semaphore.

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 7 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs]: Added reactions and polls.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 19 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs]: It's done that way now.

File: crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs (hunk 22 above)
Comment (jmartinesp) [reply crates/matrix-sdk-ui/tests/integration/timeline/pinned_event.rs]: Added a new `assert_next_matches_with_timeout!` macro for this. It's based on the `assert_next_matches`, but instead of checking the immediate result of the stream, it adds a small timeout which can be customised.

**Comment by jmartinesp** [comment]
So, funny thing: it seems like when the devs reported `/event` didn't return the sender info it was a mistake. So I'm wondering if I should go back to using `/event` for loading the pinned events since it *was* faster...

File: benchmarks/benches/room_bench.rs (hunk 25 above)
Comment (jmartinesp) [reply benchmarks/benches/room_bench.rs:193]: Tried fixing this in https://github.com/matrix-org/matrix-rust-sdk/pull/3840/commits/807c87525854af90989e9920f5d5ed07254ed60a, please let me know if there's a better solution.

File: crates/matrix-sdk/src/room/mod.rs (hunk 10 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/room/mod.rs:449]: Done in https://github.com/matrix-org/matrix-rust-sdk/pull/3840/commits/919b5458fe886ec96cab4de9fe94b66bd50c404a, added some tests.

**Comment by jmartinesp** [comment]
Ok, so I re-used this PR for holding only the part of the code related to the live updates to the pinned events timeline. It depends on a previous PR being merged and it'll have a following PR containing the semaphore/buffered futures changes.

**Comment by jmartinesp** [comment]
There is something weird happening with the CI actions, only a couple of them are running 🫤 .

**Comment by jmartinesp** [comment]
The PR this one depended on has been merged and this one has been rebased, so it's ready to review again.

**Comment by jmartinesp** [comment]
Well, apparently https://github.com/matrix-org/matrix-rust-sdk/pull/3879 just broke this PR, another rebase will be needed.

File: crates/matrix-sdk/src/event_cache/mod.rs [hunk 26]
```diff
@@ -582,6 +582,23 @@ impl RoomEventCache {
             warn!("couldn't save event without event id in the event cache");
         }
     }
+
+    /// Save some events in the event cache, for further retrieval with
+    /// [`Self::event`]. This function will save them using a single lock,
+    /// as opposed to [`Self::save_event`].
+    // This doesn't insert the event into the linked chunk. In the future there'll
+    // be no distinction between the linked chunk and the separate cache.
```
Comment (Hywan) [inline crates/matrix-sdk/src/event_cache/mod.rs:590]: Why?
Can we add a `TODO` and open an issue to track this please?

**Review by Hywan — CHANGES_REQUESTED** [review]
We are almost good!

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 26 above)
Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs:590]: The comment is just a copy of the one above it written by @bnjbvr since it follows the same logic, but I think it's probably because we lack the context of where the `LinkedChunk` containing events to save would go, or if it even makes sense to add them to one (these events can come from very different sources and may not be related enough to be added as a chunk of a timeline, if I understood how `LinkedChunk` works properly).

I guess the 'in the future' part means once we have a persistent event cache, but I'm not 100% sure.

Anyway,  can create an issue for that.

Comment (jmartinesp) [reply crates/matrix-sdk/src/event_cache/mod.rs]: @Hywan I added the TODO [here](https://github.com/matrix-org/matrix-rust-sdk/pull/3840/commits/4f57f54b7b7dfba0894aa16c5b7547a104155771) and opened the issue [here](https://github.com/matrix-org/matrix-rust-sdk/issues/3886).

File: benchmarks/benches/room_bench.rs (hunk 25 above)
Comment (Hywan) [reply benchmarks/benches/room_bench.rs:188]: Excellent, thanks!

File: crates/matrix-sdk/src/event_cache/mod.rs (hunk 26 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/mod.rs]: Excellent, thanks!

**Review by Hywan — APPROVED** [review]
Good to me, thanks! Please rebase and you can merge.
